### ISSUE TITLE
Routing is "extremely" slow using BGP, with many TCP Retransmissions, to nodes outside cluster

### ISSUE THREAD

**Issue by figaw** [cilium/cilium#34972]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

equal or higher than v1.16.0 and lower than v1.17.0

### What happened?

We experience many TCP Retransmissions when trying to access services that have advertised routes using BGP.

See attached screenshot.

### How can we reproduce the issue?

Cilium is installed with:

```shell
$ cilium install --version 1.16.1 \
  --namespace=kube-system \
  --set bgpControlPlane.enabled=true \
  --set extraConfig.nodeport-addresses="10.158.97.0/24" \
  --set ipam.mode=cluster-pool \
  --set ipam.operator.clusterPoolIPv4PodCIDRList="10.140.0.0/16" \
  --set kubeProxyReplacement=true \
  --set k8sServiceHost=${API_SERVER_IP} \
  --set k8sServicePort=${API_SERVER_PORT} \
  --set rollOutCiliumPods=true
```

BGP is configured with the following manifests:

```yaml
apiVersion: cilium.io/v2alpha1
kind: CiliumBGPAdvertisement
metadata:
  name: bgp-advertisements
  labels:
    advertise: bgp
spec:
  advertisements:
    - advertisementType: "Service"
      service:
        addresses:
          - LoadBalancerIP
      selector:
        matchExpressions:
          - { key: somekey, operator: NotIn, values: ["never-used-value"] }

---
apiVersion: cilium.io/v2alpha1
kind: CiliumBGPPeerConfig
metadata:
  name: cilium-peer
spec:
  gracefulRestart:
    enabled: true
    restartTimeSeconds: 120
  families:
    - afi: ipv4
      safi: unicast
      advertisements:
        matchLabels:
          advertise: "bgp"
---
apiVersion: cilium.io/v2alpha1
kind: CiliumBGPClusterConfig
metadata:
  name: cilium-bgp
spec:
  nodeSelector:
    matchLabels:
      bgp-policy: a
  bgpInstances:
    - name: "instance-64520"
      localASN: 64520
      peers:
        - name: "router1"
          peerASN: 64521
          peerAddress: "10.158.96.1"
          peerConfigRef:
            name: "cilium-peer"
        - name: "router2"
          peerASN: 64522
          peerAddress: "10.158.96.2"
          peerConfigRef:
            name: "cilium-peer"
```

A `CiliumLoadBalancerIPPool`

```
apiVersion: "cilium.io/v2alpha1"
kind: CiliumLoadBalancerIPPool
metadata:
  name: "my-first-pool"
spec:
  blocks:
    - cidr: "10.158.92.0/24"
```

The test is made towards an nginx deployment

```
apiVersion: v1
kind: Namespace
metadata:
  name: bgp-test
---
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: nginx
  name: nginx-deployment
  namespace: bgp-test
spec:
  volumes:
    - name: shared-data
      emptyDir: {}
  initContainers:
    - name: init-write-info
      image: alpine:latest
      command: ["/bin/sh", "-c"]
      args:
        - >
          echo "Pod Name: $(hostname)" > /usr/share/nginx/html/index.html;
          echo "Node Name: $(echo $NODE_NAME)" >> /usr/share/nginx/html/index.html;
          echo "Pod IP: $(echo $POD_IP)" >> /usr/share/nginx/html/index.html;
      env:
        - name: NODE_NAME
          valueFrom:
            fieldRef:
              fieldPath: spec.nodeName
        - name: POD_IP
          valueFrom:
            fieldRef:
              fieldPath: status.podIP
      volumeMounts:
        - name: shared-data
          mountPath: /usr/share/nginx/html
  containers:
    - name: nginx
      image: nginx:latest
      command: ["/bin/sh", "-c"]
      args:
        - |
          sed -i 's/listen       80;/listen       15762;/' /etc/nginx/conf.d/default.conf &&
          nginx -g 'daemon off;'
      ports:
        - containerPort: 15762
      volumeMounts:
        - name: shared-data
          mountPath: /usr/share/nginx/html
---
apiVersion: v1
kind: Service
metadata:
  name: nginx-service
  namespace: bgp-test
  labels:
    app: nginx
spec:
  type: LoadBalancer
  loadBalancerIP: 10.158.92.30
  selector:
    app: nginx
  ports:
    - protocol: TCP
      port: 15762
      targetPort: 15762
```

From a server not in-cluster, but on the same network: (and response comes back "fine".. just.. 6 seconds later, which feels like a lot.)

```shell
username@server3:~$ curl 10.158.92.30:15762
Pod Name: nginx-deployment-5cb6bbc45d-k8252
Node Name: cp-a-2
Pod IP: 10.140.1.92
```

From servers in the cluster it's near-instant, presumably because they get their routes from the cilium-agents.

### Cilium Version

cilium-cli: v0.16.17 compiled with go1.23.1 on linux/amd64
cilium image (default):
cilium image (stable): v1.16.1
cilium image (running): 1.16.1

### Kernel Version

Linux cp-a-2 5.15.0-122-generic #132-Ubuntu SMP Thu Aug 29 13:45:52 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Client Version: v1.29.5
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.0

### Sysdump

[cilium-sysdump-20240920-101508.zip](https://github.com/user-attachments/files/17072130/cilium-sysdump-20240920-101508.zip)

### Relevant log output

```shell
k -n kube-system exec -it cilium-dfbtl -- cilium monitor --related-to 3149
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
Listening for events on 48 CPUs with 64x4096 of shared memory
Press Ctrl-C to quit
time="2024-09-20T07:46:51Z" level=info msg="Initializing dissection cache..." subsys=monitor
-> endpoint 3149 flow 0x2e493592 , identity world->29057 state new ifindex lxc3c6ebff18583 orig-ip 10.140.2.206: 10.140.2.206:47026 -> 10.140.1.92:15762 tcp SYN
-> overlay flow 0x91bd05ed , identity 29057->remote-node state reply ifindex cilium_vxlan orig-ip 0.0.0.0: 10.140.1.92:15762 -> 10.140.2.206:47026 tcp SYN, ACK

### this is where the ~6 seconds delay happens

-> endpoint 3149 flow 0x2e493592 , identity world->29057 state established ifindex lxc3c6ebff18583 orig-ip 10.140.2.206: 10.140.2.206:47026 -> 10.140.1.92:15762 tcp ACK
-> overlay flow 0x3ae94701 , identity 29057->remote-node state reply ifindex cilium_vxlan orig-ip 0.0.0.0: 10.140.1.92:15762 -> 10.140.2.206:47026 tcp ACK
-> overlay flow 0x3ae94701 , identity 29057->remote-node state reply ifindex cilium_vxlan orig-ip 0.0.0.0: 10.140.1.92:15762 -> 10.140.2.206:47026 tcp ACK
-> endpoint 3149 flow 0x2e493592 , identity world->29057 state established ifindex lxc3c6ebff18583 orig-ip 10.140.2.206: 10.140.2.206:47026 -> 10.140.1.92:15762 tcp ACK, FIN
-> overlay flow 0x3ae94701 , identity 29057->remote-node state reply ifindex cilium_vxlan orig-ip 0.0.0.0: 10.140.1.92:15762 -> 10.140.2.206:47026 tcp ACK, FIN
-> endpoint 3149 flow 0x2e493592 , identity world->29057 state established ifindex lxc3c6ebff18583 orig-ip 10.140.2.206: 10.140.2.206:47026 -> 10.140.1.92:15762 tcp ACK
^C
Received an interrupt, disconnecting from monitor...
```

### Anything else?

Screenshot of pcap:

![image](https://github.com/user-attachments/assets/77e8e9cf-aa9b-4129-bac5-9d65200628e5)

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by figaw** [comment]
Output from `cilium bgp`

```shell
root@cp-a-2:~# cilium bgp peers
Node     Local AS   Peer AS   Peer Address   Session State   Uptime      Family         Received   Advertised
a-cp-1   64520      64521     10.158.96.1    established     17h29m45s   ipv4/unicast   1          2
         64520      64522     10.158.96.2    established     17h29m44s   ipv4/unicast   1          2
cp-a-2   64520      64521     10.158.96.1    established     17h29m34s   ipv4/unicast   1          2
         64520      64522     10.158.96.2    established     17h29m38s   ipv4/unicast   1          2
root@cp-a-2:~# cilium bgp routes
(Defaulting to `available ipv4 unicast` routes, please see help for more options)

Node     VRouter   Prefix            NextHop   Age         Attrs
a-cp-1   64520     10.158.92.30/32   0.0.0.0   17h29m52s   [{Origin: i} {Nexthop: 0.0.0.0}]
cp-a-2   64520     10.158.92.30/32   0.0.0.0   17h29m45s   [{Origin: i} {Nexthop: 0.0.0.0}]
```

**Comment by figaw** [comment]
Routes are configured identically on both peers, on the other end. Miktorik routers, RB5009UG+S+IN

(Router1 uses AS 64521, and router2 uses AS 64522)

```
[admin@a-r-1] /routing/bgp/connection> print detail
Flags: D - dynamic, X - disabled, I - inactive
 0   name="PEER_TO_a-cp-1"
     remote.address=10.158.97.5
     local.address=10.158.96.1 .role=ibgp
     routing-table=main as=64521 address-families=ip
     output.default-originate=always

 1   name="PEER_TO_2-cp-a"
     remote.address=10.158.97.7
     local.address=10.158.96.1 .role=ibgp
     routing-table=main as=64521 address-families=ip
     output.default-originate=always
```

Session on router1:

```
[admin@a-r-1] /routing/bgp/session> print detail
Flags: E - established
 0 E name="PEER_TO_2-cp-a-1"
     remote.address=10.158.97.7 .as=64521 .id=10.158.97.7 .capabilities=mp,rr,enhe,gr,as4,fqdn .afi=ip .hold-time=1m30s .messages=2109 .bytes=40104
     .gr-time=120 .gr-afi=ip .eor=ip
     local.role=ibgp .address=10.158.96.1 .as=64521 .id=192.168.88.5 .capabilities=mp,rr,gr,as4 .afi=ip .messages=2109 .bytes=40098 .eor=""
     output.procid=21 .default-originate=always
     input.procid=21 .last-notification=ffffffffffffffffffffffffffffffff0015030603 ibgp
     multihop=yes hold-time=1m30s keepalive-time=30s uptime=17h33m34s300ms last-started=2024-09-19 16:51:45 last-stopped=2024-09-19 16:17:00 prefix-count=1

 1 E name="PEER_TO_a-cp-1-1"
     remote.address=10.158.97.5 .as=64521 .id=10.158.97.5 .capabilities=mp,rr,enhe,gr,as4,fqdn .afi=ip .hold-time=1m30s .messages=2109 .bytes=40104
     .gr-time=120 .gr-afi=ip .gr-afi-fwp=ip .eor=ip
     local.role=ibgp .address=10.158.96.1 .as=64521 .id=192.168.88.5 .capabilities=mp,rr,gr,as4 .afi=ip .messages=2109 .bytes=40098 .eor=""
     output.procid=20 .default-originate=always
     input.procid=20 .last-notification=ffffffffffffffffffffffffffffffff0015030603 ibgp
     multihop=yes hold-time=1m30s keepalive-time=30s uptime=17h33m45s390ms last-started=2024-09-19 16:51:34 last-stopped=2024-09-19 16:17:15 prefix-count=1
```

Advertisements:

```
[admin@a-r-1] /routing/bgp/advertisements> print detail
 0 peer=PEER_TO_a-cp-1-1 dst=0.0.0.0/0 afi=ip local-pref=100 nexthop=<public-ip-omitted> origin=0

 0 peer=PEER_TO_2-cp-a-1 dst=0.0.0.0/0 afi=ip local-pref=100 nexthop=<public-ip-omitted> origin=0
```

Routes on router:

```
[admin@a-r-1] /ip/route> print
Flags: D - DYNAMIC; A - ACTIVE; c - CONNECT, s - STATIC, b - BGP
Columns: DST-ADDRESS, GATEWAY, DISTANCE
#     DST-ADDRESS      GATEWAY      DISTANCE
0  As 0.0.0.0/0        46.<omitted>.1         1
  D b 10.158.92.30/32  10.158.97.7       200
  DAb 10.158.92.30/32  10.158.97.5       200
  DAc 10.158.96.0/22   bridge              0
  DAc 10.158.96.3/32   vrrp-lan            0
  DAc 46.<omitted>/23   ether1              0
  DAc 46.<omitted>/32   vrrp1               0
  DAc 192.168.19.0/24  bridge              0
  DAc 192.168.88.0/24  bridge              0
```

**Comment by figaw** [comment]
We also previously had issues with the cidr of the `CiliumLoadBalancerIPPool` overlapping with the lan subnet `10.158.96.0/22`, choosing e.g. `10.158.98.0/24` for the ExternalIPs of the services. I thought this was a good idea, when I wanted the servers on the network to be able to connect to the services in the cluster. It seems I was wrong.

**Comment by brujoand** [comment]
This looks very similar to what I've been debugging for the last couple of days. Do you also see the tests for `'unexpected` packet drops' fail when running 'cilium connectivity test'?

**Comment by figaw** [comment]
Our tests are all green now, but "sometimes" when we set `bpf.masquerade=true`, the tests start failing, and setting `bpf.masquerade=false` doesn't make the problem go away, the tests stay broken until we reset the cluster.

**Comment by rastislavs** [comment]
Hi @figaw , looking at the sysdump, it seems that Cilium is advertising the routes properly as configured and the BGP peering is stable. How does the routing between the client server and the two routers in front of the cluster work? Could you ensure that there is no routing loop or asymmetric routing happening in the network?

**Comment by figaw** [comment]
Hmm, traceroute from the client does show a loop 🤔

```shell
username@client ~$ traceroute 10.158.92.30
traceroute to 10.158.92.30 (10.158.92.30), 30 hops max, 60 byte packets
 1  10.158.96.1 (10.158.96.1)  0.365 ms  0.307 ms  0.312 ms
 2  10.158.97.5 (10.158.97.5)  0.429 ms  0.380 ms  0.494 ms
 3  10.158.96.1 (10.158.96.1)  0.738 ms  0.716 ms  1.671 ms
 4  * * *
 5  * * *
 ...
 ```
 
So it seems the router is correctly forwarding the traffic to `10.158.97.5`, but that server doesn't have the service IP (`10.158.92.30`) in its routing table..

```shell
username@server ~$ route -n
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         10.158.96.3     0.0.0.0         UG    100    0        0 bond0
10.4.0.0        0.0.0.0         255.255.255.0   U     0      0        0 nerdctl0
10.4.1.0        0.0.0.0         255.255.255.0   U     0      0        0 br-8bd22b51ab86
10.140.0.0      10.140.1.98     255.255.255.0   UG    0      0        0 cilium_host
10.140.1.0      10.140.1.98     255.255.255.0   UG    0      0        0 cilium_host
10.140.1.98     0.0.0.0         255.255.255.255 UH    0      0        0 cilium_host
10.140.2.0      10.140.1.98     255.255.255.0   UG    0      0        0 cilium_host
10.158.96.0     0.0.0.0         255.255.252.0   U     100    0        0 bond0
10.158.96.3     0.0.0.0         255.255.255.255 UH    100    0        0 bond0
```

So I guess it's sending the traffic back to the router 🤷

Note: `10.158.96.3` is a VRRP ip that the routers share. When they're both online, traffic goes to `10.158.96.1`.

**Comment by harsimran-pabla** [comment]
HI @figaw, it does look like packets are getting hairpinned. 

Traceroute might be using icmp by default, and icmp to service VIP is not suppose to work in the clusters. 
https://github.com/cilium/cilium/issues/14118

Do you see TCP packets also getting hairpinned back to the router? I'd check what is the destination IP ( it would be service VIP initially, and does it get translated to one of the Pod IPs and hairpinned back to the router. )

If above is happening, you might need to set `enable-endpoint-routes: true` to have /32 pod routes installed in the kernel routing table.

**Comment by figaw** [comment]
Ah, alright. Thank you! When I run a traceroute with tcp, it returns instantly,

```shell
username@client ~$ sudo traceroute -T -p 15762 10.158.92.30
traceroute to 10.158.92.30 (10.158.92.30), 30 hops max, 60 byte packets
 1  10.158.96.1 (10.158.96.1)  0.375 ms  0.287 ms  0.325 ms
 2  * * *
 3  * * *
 4  10.158.92.30 (10.158.92.30)  0.832 ms  0.671 ms  0.656 ms
 ```

The service has a clusterIP of `10.99.69.255`, externalIP (above) `10.158.92.30`, and the pod behind it has an IP of `10.140.1.166`. The ports `15762` is the assigned one to the service, and `32570` is the nodeport. (We also use `15762` for the container-ports and bind nginx to that, just so it's somewhere else than 80 or 8080 while debugging.)

This is a tcpdump from the server, listening on all interfaces, for traffic to port `15762`.

![image](https://github.com/user-attachments/assets/3461e88e-3a27-4299-9c46-a5845169243c)

**Comment by figaw** [comment]
We tried `enable-endpoint-routes: true` with the helm flag `--set endpointRoutes.enabled=true` and the tests started failing with:

```
[=] [cilium-test-1] Test [no-unexpected-packet-drops] [1/102]
.
  [.] Action [no-unexpected-packet-drops/no-unexpected-packet-drops/cluster-a/cp-a-2]
  [-] Scenario [no-unexpected-packet-drops/no-unexpected-packet-drops]
  [.] Action [no-unexpected-packet-drops/no-unexpected-packet-drops/cluster-a/mw1]
  ❌ Found unexpected packet drops:
{
  "labels": {
    "direction": "INGRESS",
    "reason": "FIB lookup failed"
  },
  "name": "cilium_drop_count_total",
  "value": 144
}
{
  "labels": {
    "direction": "INGRESS",
    "reason": "First logical datagram fragment not found"
  },
  "name": "cilium_drop_count_total",
  "value": 1
}
.  ❌ Found unexpected packet drops:
{
  "labels": {
    "direction": "INGRESS",
    "reason": "TTL exceeded"
  },
  "name": "cilium_drop_count_total",
  "value": 12
}
  [.] Action [no-unexpected-packet-drops/no-unexpected-packet-drops/cluster-a/a-cp-1]
.  ❌ Found unexpected packet drops:
{
  "labels": {
    "direction": "EGRESS",
    "reason": "TTL exceeded"
  },
  "name": "cilium_drop_count_total",
  "value": 12
}
```

We tried to give it 30 and 60 minutes to settle in, but to no avail.

**Comment by rtrox** [comment]
I've been beating my head against a similar issue for months, with almost the same behavior -- 6-7 second delay on initial TCP connection, and tcpdump shows numerous retransmits during that delay. Also with BGP driven routing. Turns out I had a dumb firewall issue, sharing here in the off chance that you, or someone else who stumbles on this thread had the same problem.

The initial connection to a node from a client outside the cluster is routed through the router via BGP, but the router uses an ICMP redirect to share the route with the client, which is somehow leading to an invalid connection state. I had a rule pretty high up in my router's firewall chain to drop all invalid connections, which was causing the connection to be blocked at the firewall until it was reinitiated directly from the client to the node. Just adding an exception to allow "invalid" connection states to my LoadBalancer IP Range was enough to fix the delay.

If you're curious how I got there -- your comment that you thought static routes were why the nodes had no issues inter-communicating was what tipped me off. How I got there, in case you're smarter than me, but the steps help you find your answer :-P
1. On a hunch from your comment, I added a static route to my laptop for the LB IP Range, to cut out the router. This instantly fixed the delay, so I knew the issue was confined to when traffic was actually being routed by the router.
2. To rule out the firewall, I added a blanket allow rule for the LB IPs to my router, and again, instantly fixed the delay.
3. moved that rule down the firewall chain until the delay returned to isolate the bad rule.

**Comment by figaw** [comment]
Thank you @rtrox! Thank you so much! The delay is gone.

We tried with the static routes as well from the LB cidr to the nodes in the cluster, and the connection was instant for all of them, which was a good test.

We then tried to add the firewall rule, we had a default "drop invalid" in our forward chain, and when we allowed invalid requests to the LB cidr above it, the connection worked!

We still have the tests failing, but it feels like those might be related to other things.

If someone wants to elaborate on why allowing `connection-state: invalid` for the load-balancer cidr is necessary, that would be much appreciated. 

I tried to disable "send redirects" on our router, `IP -> Settings -> Send Redirects`, but that didn't change anything; the firewall rule was still necessary, so I don't know what the "proper way" to do this is. Having an allow-rule for invalid packages seems  undesirable. <https://help.mikrotik.com/docs/spaces/ROS/pages/103841817/IP+Settings>

**Comment by jarulsamy** [comment]
> I've been beating my head against a similar issue for months, with almost the same behavior -- 6-7 second delay on initial TCP connection, and tcpdump shows numerous retransmits during that delay. Also with BGP driven routing. Turns out I had a dumb firewall issue, sharing here in the off chance that you, or someone else who stumbles on this thread had the same problem.
> 
> The initial connection to a node from a client outside the cluster is routed through the router via BGP, but the router uses an ICMP redirect to share the route with the client, which is somehow leading to an invalid connection state. I had a rule pretty high up in my router's firewall chain to drop all invalid connections, which was causing the connection to be blocked at the firewall until the client timed out the connection and reinitiated it directly to the node. Just adding an exception to allow "invalid" connection states to my LoadBalancer IP Range was enough to fix the delay.
> 
> If you're curious how I got there -- your comment that you thought static routes were why the nodes had no issues inter-communicating was what tipped me off. How I got there, in case you're smarter than me, but the steps help you find your answer :-P
> 
>     1. On a hunch from your comment, I added a static route to my laptop for the LB IP Range, to cut out the router. This instantly fixed the delay, so I knew the issue was confined to when traffic was actually being routed by the router.
> 
>     2. To rule out the firewall, I added a blanket allow rule for the LB IPs to my router, and again, instantly fixed the delay.
> 
>     3. moved that rule down the firewall chain until the delay returned to isolate the bad rule.

Seriously, thank you so much!. I dealt with a very similar problem with BGP on my Opnsense router. Turns out I needed an outbound NAT rule (https://forum.opnsense.org/index.php?topic=27063.0) to ensure traffic flowed symmetrically through my firewall. Not exactly the same issue, but your solution led me the right way. Sorry to take the thread off topic, but this has solved a major, _major_ problem for me.

**Comment by figaw** [comment]
This is still relevant, either as a bug or documentation.

**Comment by kiemlicz** [comment]
I'm wondering whether this issue is connected to Cilium or rather Mikrotiks or any other routing performing something like described on [Mikrotik forum (check Solution post)](https://forum.mikrotik.com/viewtopic.php?t=171177)

As I try to understand this: the reply from server might omit the L3 connection tracking and get dropped (?) if router sees the network it sends the reply to as directly attached

**Comment by likeMyCoffee** [comment]
I don’t use Cilium, but I think I know the issue because it looks a lot like a BGP side-effect we have in kubevip. Took me more than a week to figure out…

What could be happening is that you are connecting to the BGP service from the kubernetes subnet containing the load balanced servers. This results in asynchronous routing.

Something like the following happens:
You open a session to the service and it loadbalances to a kubernetes node.
Because your IP is in the same network as the system you are connecting with, the returning packets go directly back (no longer via router).
Your system rejects the now new incoming connection.
No reply is received and your system starts retransmitting.
The router also only sees half of the connections.

The simple solution is to never connect from the internal kubernetes subnet to a load balanced service. The kubernetes environment response should always need to be send to the router.

Hope this is the issue and above helps

**Comment by figaw** [comment]
This is still relevant, either as a bug or documentation.

**Comment by squeed** [comment]
@figaw noted. Could you summarize the *current* issue? (I only ask as there's quite a complicated comment history that is not the easiest to grok.) That should make it easier to determine if it's a bug or documentation gap.

**Comment by matejaputic** [comment]
Adding this in case it helps someone. After banging my head for a few days over this, I finally found a solution for my scenario. After adding a static route to the cluster from my local host that bypasses the router, I narrowed it down to my router's firewall.

TL;DR:
```
/ip/firewall/add action=accept chain=forward comment="Allow BGP routes without FastTrack" dst-address=<cilium-bgp-ippool>
```
This rule must be placed before the `defconf: fasttrack` rule.

My setup:
- MikroTik RouterOS 7.18.2
- Cilium 1.18.0-pre.0 on Talos 1.9.5
- Simplest possible network topology: localhost -- router -- cluster

Apparently what was happening is that BGP packets were being FastTracked, skipping routing decisions that would normally direct them according to the BGP-learned routes. So adding the above firewall rule to process those packets normally fixes the issue for me.

I have posted my configs here: https://gist.github.com/matejaputic/bfa4fd5ac494f01a8a0b60180f77b9e2

**Comment by likeMyCoffee** [comment]
I might have a better solution for you. This worked for me in my similar situation.

On the router add an outgoing source NAT for you r cluster network range that changes the source IP to the default gateway.

So what does this do?
When you access your service from your client (which is in the same network), your BGP sends it to the service.
When the service replies the client is still in the same network but we SNAT'ed it to the default gateway.
Because the gateway is now the source, replies get sent to the gateway which is your router.
The router then relays it back to the client.

By doing this the router sees the data going both ways and communication is symmetric and stateful.

This way you don't have to modify all your clients.

**Comment by figaw** [comment]
@squeed Thanks for reaching out! The current issue is that it's unclear what requirements there is for the BGP traffic to flow properly, see this comment: https://github.com/cilium/cilium/issues/34972#issuecomment-2429601549

.. and for the two comments above, it now seems like we're getting a lot of solutions, which is great, but an opinionated solution or some guidelines should probably live in a more official location. (Hence why I suggested it's less of a bug and maybe a documentation issue... if the proper way to solve it, is to configure your firewall differently.)

.. it threw me off that.. "... allowing `connection-state: invalid` for the load-balancer cidr..." suddenly made things work 😅
